tests/loggen/socket_plugin/socket_plugin.c

```diff
@@ -0,0 +1,409 @@
+/*
+ * Copyright (c) 2007-2018 Balabit
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#define GLIB_DISABLE_DEPRECATION_WARNINGS 1
+
+#include <signal.h>
+#include <stdio.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <string.h>
+#include <glib.h>
+#include <gmodule.h>
+#include <netdb.h>
+
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <sys/un.h>
+#include <stdlib.h>
+#include <errno.h>
+
+#include "loggen_plugin.h"
+#include "loggen.h"
+#include "loggen_helper.h"
+
+static void           start(plugin_option *option);
+static void           stop(plugin_option *option);
+static gpointer       active_thread_func(gpointer user_data);
+static gpointer       idle_thread_func(gpointer user_data);
+static ssize_t        send_plain(int fd, void *buf, size_t length);
+static gint           get_thread_count(void);
+
+static GOptionEntry  *get_options(void);
+static GPtrArray      *thread_array = NULL;
+
+static gboolean thread_run;
+
+static GMutex *thread_lock = NULL;
+static GCond *thread_start = NULL;
+static GCond *thread_connected = NULL;
+static gint connect_finished;
+static gint active_thread_count;
+static gint idle_thread_count;
+
+static int inet_socket_i = 0;
+static int unix_socket_x = 0;
+static int sock_type_s = 0;
+static int sock_type_d = 0;
+
+static GOptionEntry loggen_options[] =
+{
+  { "inet", 'i', 0,   G_OPTION_ARG_NONE, &inet_socket_i, "Use IP-based transport (TCP, UDP)", NULL },
+  { "unix", 'x', 0,   G_OPTION_ARG_NONE, &unix_socket_x, "Use UNIX domain socket transport", NULL },
+  { "stream", 'S', 0, G_OPTION_ARG_NONE, &sock_type_s,   "Use stream socket (TCP and unix-stream)", NULL },
+  { "dgram", 'D', 0,  G_OPTION_ARG_NONE, &sock_type_d,   "Use datagram socket (UDP and unix-dgram)", NULL },
+  { NULL }
+};
+
+plugin_info loggen_plugin_info =
+{
+  .name = "socket-plugin",
+  .get_options_list = get_options,
+  .start_plugin = start,
+  .stop_plugin = stop,
+  .get_thread_count = get_thread_count,
+  .require_framing = FALSE
+};
+
+static gint get_thread_count(void)
+{
+  int num;
+  g_mutex_lock(thread_lock);
+  num = active_thread_count + idle_thread_count;
```

Comment:
Why do we need to lock here? As I see active_thread_count is only decremented, none of the other instances are locked, and this get_thread_count also called periodically, so we do not miss anything if we return a wrong value. I think there is no need to lock here.

Reply:
Yes, the counters are decremented only (at the end of a thread). Because threads are finish almost the same time (the do the same job) so at a given time many threads will decrements the counter which may corrupt the value.
A corrupt value may disturb the main program as it is calling the get_thread_count() continously to check a plugin's state.